### ISSUE TITLE
DeepEP 'Test failed: Failed: Assertion error /opt/DeepEP/csrc/deep_ep.cpp:510 'x.dim() == 2 and x.is_contiguous()''

### ISSUE THREAD

**Issue by goldhuang** [pytorch/torchtitan#2242]
### Bug description

When run code below, I got error 'Test failed: Failed: Assertion error /opt/DeepEP/csrc/deep_ep.cpp:510 'x.dim() == 2 and x.is_contiguous()''
at https://github.com/pytorch/torchtitan/blob/5ef90faf24f2da07a93c286aa3727177aaf94257/torchtitan/distributed/deepep/deepep.py#L185

```
from __future__ import annotations

import gc
import torch
import torch.distributed as dist
from torch.distributed.tensor import DTensor
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor.parallel import parallelize_module
import DeepEPExpertParallel

from moe import MoEArgs, build_moe

def cleanup_cuda() -> None:
    gc.collect()  # Python garbage collection
    torch.cuda.empty_cache()  # Releases unused memory to CUDA
    torch.cuda.ipc_collect()  # Optional: collects inter-process shared memory
    return None

def run_tests(
    dim: int,
    n_experts: int,
    n_act: int,
    num_tokens: int,
) -> None:
    cleanup_cuda()
    
    dist.init_process_group(backend="nccl")
    torch.cuda.set_device(dist.get_rank()%8)
        
    ep_degree = dist.get_world_size()
    device_mesh = init_device_mesh("cuda", (ep_degree,))

    rank = dist.get_rank()
    torch.manual_seed(42+rank)

    if rank == 0:
        print("Running test_moe")
    # ------------------------------------------------------------------#
    # 1. Build configs
    # ------------------------------------------------------------------#

    args = MoEArgs(num_experts=n_experts, num_shared_experts=0, score_before_experts=False, top_k=n_act)
    baseline_moe = build_moe(args=args, dim=dim, hidden_dim=2*dim)
    ep_moe = build_moe(args=args, dim=dim, hidden_dim=2*dim, moe_impl="deepep")

    device = "cuda"
    baseline_moe, ep_moe = [
        m.to(device).bfloat16() for m in (baseline_moe, ep_moe)
    ]

    canonical_state_dict = baseline_moe.state_dict()
    for param in canonical_state_dict.values():
        dist.broadcast(param.data, src=0)

    baseline_moe.load_state_dict(canonical_state_dict)
    ep_moe.load_state_dict(canonical_state_dict)

    # Apply EP
    ep_style = DeepEPExpertParallel()
    # ep_moe = ep_style._apply(ep_moe, device_mesh)
    parallelize_module(
        module=ep_moe.experts,
        device_mesh=device_mesh,
        parallelize_plan=ep_style,
    )

    # ------------------------------------------------------------------#
    # 3. GPU + BF16 + forward -> backward
    # ------------------------------------------------------------------#

    x_baseline = torch.randn(1, num_tokens, dim, device=device, dtype=torch.bfloat16, requires_grad=True)
    
    x_ep = x_baseline.detach().clone().requires_grad_(True)
    out_ep = ep_moe(x_ep)
    out_ep.sum().backward()
            
if __name__ == "__main__":
    try:
        # run this with torchrun --nproc_per_node=4 test_ep.py
        dim = 1024
        multiple_of = 512
        moe_scale = 0.5
        n_experts = 64
        n_act = 7
        num_tokens = 20000
        run_tests(
            dim=dim,
            multiple_of=multiple_of,
            moe_scale=moe_scale,
            n_experts=n_experts,
            n_act=n_act,
            num_tokens=num_tokens,
        )
    except Exception as e:
        print(f"Test failed: {e}")
        raise e
    finally:
        dist.destroy_process_group()
```

### Versions

Latest main of torchtitan
PyTorch 2.9.0+cu130

**Comment by tianyu-l** [comment]
Do you also hit this issue when running E2E?

I wouldn't expect the code runs with arbitrary unit test. Could you help debug?

**Comment by goldhuang** [comment]
Hi team, let me add more details.

My env passed `test_intranode.py`
```
[config] num_tokens=4096, hidden=7168, num_topk=8
[layout] Kernel performance: 0.036 ms
[testing] Running with BF16, without top-k (async=False, previous=False) ... passed
[testing] Running with BF16, with top-k (async=False, previous=False) ... passed
[testing] Running with BF16, without top-k (async=False, previous=False) ... passed
[testing] Running with BF16, with top-k (async=False, previous=False) ... passed
[testing] Running with FP8, without top-k (async=False, previous=False) ... passed
[testing] Running with FP8, with top-k (async=False, previous=False) ... passed
[testing] Running with BF16, without top-k (async=True, previous=False) ... passed
[testing] Running with BF16, with top-k (async=True, previous=False) ... passed
[testing] Running with BF16, without top-k (async=True, previous=False) ... passed
[testing] Running with BF16, with top-k (async=True, previous=False) ... passed
[testing] Running with FP8, without top-k (async=True, previous=False) ... passed
[testing] Running with FP8, with top-k (async=True, previous=False) ... passed
[testing] Running with BF16, without top-k (async=False, previous=True) ... passed
[testing] Running with BF16, with top-k (async=False, previous=True) ... passed
[testing] Running with BF16, without top-k (async=False, previous=True) ... passed
[testing] Running with BF16, with top-k (async=False, previous=True) ... passed
[testing] Running with FP8, without top-k (async=False, previous=True) ... passed
[testing] Running with FP8, with top-k (async=False, previous=True) ... passed
[testing] Running with BF16, without top-k (async=True, previous=True) ... passed
[testing] Running with BF16, with top-k (async=True, previous=True) ... passed
[testing] Running with BF16, without top-k (async=True, previous=True) ... passed
[testing] Running with BF16, with top-k (async=True, previous=True) ... passed
[testing] Running with FP8, without top-k (async=True, previous=True) ... passed
[testing] Running with FP8, with top-k (async=True, previous=True) ... passed
[tuning] SMs 24, NVL chunk 4: 161.46 GB/s (NVL), 676.74 us
[tuning] SMs 24, NVL chunk 6: 234.00 GB/s (NVL), 466.95 us
[tuning] SMs 24, NVL chunk 8: 246.67 GB/s (NVL), 442.97 us
[tuning] SMs 24, NVL chunk 10: 245.17 GB/s (NVL), 445.69 us
[tuning] SMs 24, NVL chunk 12: 244.40 GB/s (NVL), 447.08 us
[tuning] SMs 24, NVL chunk 14: 239.96 GB/s (NVL), 455.35 us
[tuning] SMs 24, NVL chunk 16: 240.04 GB/s (NVL), 455.20 us
[tuning] SMs 24, NVL chunk 18: 239.39 GB/s (NVL), 456.45 us
[tuning] SMs 24, NVL chunk 20: 233.64 GB/s (NVL), 467.68 us
[tuning] SMs 24, NVL chunk 22: 235.16 GB/s (NVL), 464.66 us
[tuning] SMs 24, NVL chunk 24: 235.69 GB/s (NVL), 463.61 us
[tuning] SMs 24, NVL chunk 26: 230.14 GB/s (NVL), 474.78 us
[tuning] SMs 24, NVL chunk 28: 230.22 GB/s (NVL), 474.62 us
[tuning] SMs 24, NVL chunk 30: 232.08 GB/s (NVL), 470.82 us
[tuning] SMs 24, NVL chunk 32: 225.44 GB/s (NVL), 484.68 us
[tuning] SMs 24, NVL chunk default: 233.63 GB/s (NVL), 467.69 us
[tuning] Best dispatch (FP8): SMs 24, NVL chunk 8, 246.67 GB/s (NVL), t: 442.97 us
[tuning] SMs 24, NVL chunk 4: 258.67 GB/s (NVL), 819.24 us
[tuning] SMs 24, NVL chunk 6: 256.74 GB/s (NVL), 825.42 us
[tuning] SMs 24, NVL chunk 8: 255.83 GB/s (NVL), 828.36 us
[tuning] SMs 24, NVL chunk 10: 254.51 GB/s (NVL), 832.63 us
[tuning] SMs 24, NVL chunk 12: 253.55 GB/s (NVL), 835.80 us
[tuning] SMs 24, NVL chunk 14: 253.03 GB/s (NVL), 837.50 us
[tuning] SMs 24, NVL chunk 16: 251.52 GB/s (NVL), 842.55 us
[tuning] SMs 24, NVL chunk 18: 250.03 GB/s (NVL), 847.55 us
[tuning] SMs 24, NVL chunk 20: 250.22 GB/s (NVL), 846.92 us
[tuning] SMs 24, NVL chunk 22: 249.19 GB/s (NVL), 850.41 us
[tuning] SMs 24, NVL chunk 24: 247.88 GB/s (NVL), 854.90 us
[tuning] SMs 24, NVL chunk 26: 247.06 GB/s (NVL), 857.74 us
[tuning] SMs 24, NVL chunk 28: 246.36 GB/s (NVL), 860.17 us
[tuning] SMs 24, NVL chunk 30: 244.91 GB/s (NVL), 865.28 us
[tuning] SMs 24, NVL chunk 32: 244.19 GB/s (NVL), 867.82 us
[tuning] SMs 24, NVL chunk default: 257.01 GB/s (NVL), 824.53 us
[tuning] Best dispatch (BF16): SMs 24, NVL chunk 4, 258.67 GB/s (NVL), t: 819.24 us
[tuning] SMs 24, NVL chunk 1: 66.98 GB/s (NVL), 3163.74 us
[tuning] SMs 24, NVL chunk 2: 134.91 GB/s (NVL), 1570.78 us
[tuning] SMs 24, NVL chunk 3: 198.50 GB/s (NVL), 1067.55 us
[tuning] SMs 24, NVL chunk 4: 255.29 GB/s (NVL), 830.08 us
[tuning] SMs 24, NVL chunk 5: 305.05 GB/s (NVL), 694.69 us
[tuning] SMs 24, NVL chunk 6: 343.49 GB/s (NVL), 616.94 us
[tuning] SMs 24, NVL chunk 7: 303.81 GB/s (NVL), 697.53 us
[tuning] SMs 24, NVL chunk 8: 328.55 GB/s (NVL), 644.99 us
[tuning] SMs 24, NVL chunk 9: 349.06 GB/s (NVL), 607.10 us
[tuning] SMs 24, NVL chunk 10: 353.05 GB/s (NVL), 600.23 us
[tuning] SMs 24, NVL chunk 11: 348.02 GB/s (NVL), 608.92 us
[tuning] SMs 24, NVL chunk 12: 348.93 GB/s (NVL), 607.33 us
[tuning] SMs 24, NVL chunk 13: 336.69 GB/s (NVL), 629.40 us
[tuning] SMs 24, NVL chunk 14: 343.69 GB/s (NVL), 616.58 us
[tuning] SMs 24, NVL chunk 15: 344.37 GB/s (NVL), 615.37 us
[tuning] SMs 24, NVL chunk 16: 341.14 GB/s (NVL), 621.20 us
[tuning] SMs 24, NVL chunk default: 348.23 GB/s (NVL), 608.55 us
[tuning] Best combine: SMs 24, NVL chunk 10: 353.05 GB/s (NVL), t: 600.23 us
```

I'm using 4 GPUs to do the test with ep_size=4, with MoE layer in torchtitan. For the forward, the shape after _token_dispatch() and _token_combine() are correct.
```
# topK = 7 and seqlen per GPU is 20000
Rank 0 - dispatched 141265
Rank 3 - dispatched 139154
Rank 2 - dispatched 139612
Rank 1 - dispatched 139969
Rank 0 - combined 20000
Rank 3 - combined 20000
Rank 2 - combined 20000
Rank 1 - combined 20000
```
But the `routed_output` after _token_combine() is with all 0s. (hidden_status after _token_dispatch() is correct.)

Could you share any insight on this issue? Is there any hardcoded limitation in the DeepEP implementation so that it only work under certain assumption?

Thanks a lot!

**Comment by tianyu-l** [comment]
@elfiegg I wonder if you have insights

**Comment by elfiegg** [comment]
Did I understand correctly the issue is when you run EP_size=4 and GPU_size=4, your outputs after combine would become all 0s? 

The hardcode limitation of DeepEP is it only supports NVL8, GPUs via NVLink greater than 8 won't run; But it should support your case. 

Becoming all 0s sounds like your outputs are multiplied with a specifically initialized all-0s tensor (otherwise the outputs would be random). Did you check your topk_weights(topk_scores)? 

A reproducer would always be appreciated :)

**Comment by goldhuang** [comment]
Thanks! @elfiegg @tianyu-l 
I managed to fix the issue and the test code attached in the thread pass now.

The only input from my side is that `'Test failed: Failed: Assertion error /opt/DeepEP/csrc/deep_ep.cpp:510 'x.dim() == 2 and x.is_contiguous()''` could happen when x is not contiguous, because the C++ code requires contiguous x.
